test/parallel/test-timers-promisified.js

```diff
@@ -152,16 +295,99 @@ process.on('multipleResolves', common.mustNotCall());
 
 {
   exec(`${process.execPath} -pe "const assert = require('assert');` +
-       'require(\'timers/promises\').setTimeout(1000, null, { ref: false }).' +
-       'then(assert.fail)"').then(common.mustCall(({ stderr }) => {
+    'require(\'timers/promises\').setTimeout(1000, null, { ref: false }).' +
+    'then(assert.fail)"').then(common.mustCall(({ stderr }) => {
+    assert.strictEqual(stderr, '');
+  }));
+}
+
+{
+  exec(`${process.execPath} -pe "const assert = require('assert');` +
+    'require(\'timers/promises\').setImmediate(null, { ref: false }).' +
+    'then(assert.fail)"').then(common.mustCall(({ stderr }) => {
     assert.strictEqual(stderr, '');
   }));
 }
 
 {
   exec(`${process.execPath} -pe "const assert = require('assert');` +
-       'require(\'timers/promises\').setImmediate(null, { ref: false }).' +
-       'then(assert.fail)"').then(common.mustCall(({ stderr }) => {
+    'const interval = require(\'timers/promises\')' +
+    '.setInterval(1000, null, { ref: false });' +
+    'interval[Symbol.asyncIterator]().next()' +
+    '.then(assert.fail)"').then(common.mustCall(({ stderr }) => {
     assert.strictEqual(stderr, '');
   }));
 }
+
+{
+  async function runInterval(fn, intervalTime, signal) {
+    const input = 'foobar';
+    const interval = setInterval(intervalTime, input, { signal });
+    let iteration = 0;
+    for await (const value of interval) {
+      assert.strictEqual(value, input);
+      iteration++;
+      await fn(iteration);
+    }
+  }
+
+  {
+    // Check that we call the correct amount of times.
+    const controller = new AbortController();
+    const { signal } = controller;
+
+    let loopCount = 0;
+    const delay = 20;
+    const timeoutLoop = runInterval(() => {
+      loopCount++;
+      if (loopCount === 5) controller.abort();
+      if (loopCount > 5) throw new Error('ran too many times');
+    }, delay, signal);
+
+    assert.rejects(timeoutLoop, /AbortError/).then(common.mustCall(() => {
+      assert.strictEqual(loopCount, 5);
+    }));
+  }
+
+  {
+    // Check that if we abort when we have some callbacks left,
+    // we actually call them.
+    const controller = new AbortController();
+    const { signal } = controller;
+    const delay = 10;
+    let totalIterations = 0;
+    const timeoutLoop = runInterval(async (iterationNumber) => {
+      if (iterationNumber === 2) {
+        await setTimeout(delay * 2);
+        controller.abort();
+      }
+      if (iterationNumber > totalIterations) {
+        totalIterations = iterationNumber;
+      }
+    }, delay, signal);
+
+    timeoutLoop.catch(common.mustCall(() => {
+      assert.ok(totalIterations >= 3, `iterations was ${totalIterations} < 3`);
+    }));
+  }
+}
```

Comment:
Unfortunately, this test is unreliable and is failing in CI on  Raspberry Pi devices. I also can make it fail trivially on my macOS laptop by running `tools/test.py -j96 --repeat=192 test/parallel/test-timers-promisified.js`. I haven't looked closely (yet) but in my experience, having a magic number `delay` variable like this in a timers test indicates an assumption that the host isn't so slow that (say) a 10ms second timer won't fire 50ms later. This assumption is, of course, incorrect if the machine has low CPU/memory (like a Raspberry Pi device) or if there are a lot of other things competing for the machine's resources (`-j96 --repeat=192`).

Reply:
Sorry for causing an issue - I'll take a look and fix it

Reply:
When implementing this, I was under the assumption that if we I have an interval and a timeout where the interval starts before a timeout and is supposed to execute before the timeout executes, the interval's callback would always execute before the timeout's callback. Either this assumption was wrong, or something in my setInterval implementation is incorrect.

Reply:
https://github.com/nodejs/node/pull/37227